engine/src/main/java/com/sandboni/core/engine/filter/ChangeScopeFilter.java

```diff
@@ -0,0 +1,33 @@
+package com.sandboni.core.engine.filter;
+
+import com.sandboni.core.scm.scope.Change;
+import com.sandboni.core.scm.scope.ChangeScope;
+
+import java.io.File;
+import java.util.List;
+import java.util.Set;
+
+import static java.util.stream.Collectors.*;
+
+public class ChangeScopeFilter implements ScopeFilter<ChangeScope<Change>, Set<File>> {
+
+    private static final String JAVA_EXT = ".java";
+    private static final String CLASS_EXT = ".class";
+
+    @Override
+    public boolean isInScope(ChangeScope<Change> changeScope, Set<File> mainSourceDirs) {
+        // cartesian product of mainSourceDirs and changeScope
+        List<String> filePaths = mainSourceDirs.stream()
+                .map(dir -> changeScope.getAllAffectedClasses().stream()
+                        .map(className -> className.replace(JAVA_EXT, CLASS_EXT))
+                        .map(name -> dir + File.separator + name)
+                        .collect(toList()))
+                .flatMap(List::stream)
+                .collect(toList());
+
+        // we just need to find one changed class in the current module source code
+        return filePaths.stream()
+                .map(File::new)
```

Comment:
Performance implications?

Reply:
Much less than scan the complete module if the changes are not located here. Also, one file is created at a time and when we find the first one, we stop the computation of all additional files.